### PR TITLE
stacks: emit events for deferred actions

### PR DESCRIPTION
**Pull request by liamcervante**
This PR updates the stacks runtime so it (a) emits a dedicated event for deferred resources and (b) includes the deferred resources in the per-component resource counts.

There's a small refactoring to reuse the `Deferred` definition between the stacks raw data and the internal Terraform plan. This is copying the pattern used elsewhere in the stacks raw data, where identical messages are just pulled directly from the plan.

### CODE REVIEW

File: internal/rpcapi/stacks.go [hunk 1]
```diff
@@ -34,6 +36,21 @@ type stacksServer struct {
 	stopper            *stopper
 	handles            *handleTable
 	experimentsAllowed bool
+
+	// providerCacheOverride is a map of provider names to provider factories
+	// that should be used instead of the default provider cache. This is used
+	// within tests to side load providers without needing a real provider
+	//	cache.
```
Comment (DanielMSchmidt) [inline internal/rpcapi/stacks.go:43]:
```suggestion
	// cache.
```

File: internal/rpcapi/stacks.go [hunk 2]
```diff
@@ -765,42 +790,42 @@ func stackChangeHooks(send func(*terraform1.StackChangeProgress) error, mainStac
...
+			send(&terraform1.StackChangeProgress{
+				Event: &terraform1.StackChangeProgress_ResourceInstancePlannedChange_{
+					ResourceInstancePlannedChange: ripc,
+				},
+			})
+			return span
+		},
+
+		ReportResourceInstanceDeferred: func(ctx context.Context, span any, change *hooks.DeferredResourceInstanceChange) any {
+			span.(trace.Span).AddEvent("deferred resource instance", trace.WithAttributes(
```
Comment (DanielMSchmidt) [inline internal/rpcapi/stacks.go:812]: Just out of curiosity, why are we tracing this but not the planned?

**Review by DanielMSchmidt — APPROVED** [review]
Just as a nit, splitting the refactoring and new functionality up into two commits would have made reviewing a bit easier. Still ok at that size though :)

File: internal/rpcapi/stacks.go (hunk 2 above)
Comment (liamcervante) [reply internal/rpcapi/stacks.go:802]: The planned should be traced on line 783.

File: internal/rpcapi/stacks.go [hunk 3]
```diff
@@ -765,42 +780,42 @@ func stackChangeHooks(send func(*terraform1.StackChangeProgress) error, mainStac
 		// Upon completion of a component instance plan, we emit a planned
 		// change sumary event to the client for each resource instance.
 		ReportResourceInstancePlanned: func(ctx context.Context, span any, ric *hooks.ResourceInstanceChange) any {
-			actions, err := terraform1.ChangeTypesForPlanAction(ric.Change.Action)
-			if err != nil {
-				// TODO: what do we do?
-				return span
-			}
-
 			span.(trace.Span).AddEvent("planned resource instance", trace.WithAttributes(
```
Comment (liamcervante) [inline internal/rpcapi/stacks.go:783]: @DanielMSchmidt, here!

**Pull request merged**